Sources/PackageGraph/PackageGraph+Loading.swift

```diff
@@ -471,14 +410,16 @@ private func createResolvedPackages(
                 return false
             })
 
+        let lookupByProductIDs = packageBuilder.package.manifest.toolsVersion >= .vNext || moduleAliasingUsed
+
         // Get all the products from dependencies of this package.
         let productDependencies = packageBuilder.dependencies
             .flatMap({ (dependency: ResolvedPackageBuilder) -> [ResolvedProductBuilder] in
                 // Filter out synthesized products such as tests and implicit executables.
-                let explicit = Set(dependency.package.manifest.products.lazy.map({ $0.name }))
-                return dependency.products.filter({ explicit.contains($0.product.name) })
+                let explicit = Set(dependency.package.products.lazy.map({ lookupByProductIDs ?  $0.ID : $0.name }))
```

Comment:
nit: I know it's pre-existing but it might be worth to have a follow-up PR to replace `ID` with `identity`, at least it always trips me up and I have to lookup again that this is essentially using package identity + product name. Not sure if others feel the same.